oak_functions/loader/src/server.rs

```diff
@@ -14,65 +14,207 @@
 // limitations under the License.
 
 use anyhow::Context;
+use byteorder::{ByteOrder, LittleEndian};
 use http::{request::Request, response::Response};
 use hyper::{
     service::{make_service_fn, service_fn},
     Body, Server, StatusCode,
 };
-use log::info;
+use log::{error, info};
+use oak_functions_abi::OakStatus;
 use std::sync::Arc;
+use wasmi::ValueType;
 
 const MAIN_FUNCTION_NAME: &str = "main";
 
-// An ephemeral request handler with a Wasm module for handling the requests.
-#[derive(Clone)]
-pub(crate) struct WasmHandler {
-    // Wasm module to be served on each invocation. `Arc` is needed to make `WasmHandler`
-    // cloneable.
-    module: Arc<wasmi::Module>,
+/// Wasm host function index numbers for `wasmi` to map import names with. This numbering is not
+/// exposed to the Wasm client.  See https://docs.rs/wasmi/0.6.2/wasmi/trait.Externals.html
+const READ_REQUEST: usize = 0;
+const WRITE_RESPONSE: usize = 1;
+
+// Type aliases for positions and offsets in Wasm linear memory. Any future 64-bit version
+// of Wasm would use different types.
+type AbiPointer = u32;
+type AbiPointerOffset = u32;
+/// Wasm type identifier for position/offset values in linear memory. Any future 64-bit version of
+/// Wasm would use a different value.
+const ABI_USIZE: ValueType = ValueType::I32;
+
+/// `WasmInterface` holds runtime values for a particular execution instance of Wasm, handling a
+/// single user request. The methods here correspond to the ABI host functions that allow the Wasm
+/// module to exchange the request and the response with the Oak functions server. These functions
+/// translate values between Wasm linear memory and Rust types.
+struct WasmInterface {
+    request_bytes: Vec<u8>,
+    response_bytes: Vec<u8>,
+    memory: Option<wasmi::MemoryRef>,
```

Comment:
This Optional field and the two-step construction process does not seem very elegant, I hope we can make things more streamlined in the future :)